In [1]:
import pandas as pd
from morfeusz2 import Morfeusz
morfeusz = Morfeusz() 

In [2]:
pd.set_option('display.max_colwidth', None)

## Loading data and choosing only appropiate MWE

In [3]:
poleval_df = pd.read_csv("poleval2019_task2_training_190221/index.tsv", sep="\t", header=None)
poleval_df.columns = ["phrase_id", "doc_id", "text", "lemma"]
poleval_df.head()

phrase_id  doc_id                          text  \
0      31822   99883       Toronto Dominion Centre   
1      40025   99883                       Toronto   
2     343873   99883  kompleks handlowo-kulturalny   
3      31833   99883                    Joe Fafard   
4     327365   99883                   kanadyjskim   

                          lemma  
0       Toronto Dominion Centre  
1                       Toronto  
2  kompleks handlowo-kulturalny  
3                    Joe Fafard  
4                    kanadyjski

In [4]:
print(f"Shape of the original Poleval 2019 dataset: {poleval_df.shape}")

Shape of the original Poleval 2019 dataset: (22177, 4)


In [5]:
mwe_df = poleval_df[poleval_df["text"].str.split().str.len() > 1]
mwe_df.head()

phrase_id  doc_id                                   text  \
0      31822   99883                Toronto Dominion Centre   
2     343873   99883           kompleks handlowo-kulturalny   
3      31833   99883                             Joe Fafard   
5     343872   99883              Ludwiga Mies van der Rohe   
7     343871   99883  Toronto Dominion Gallery of Inuit Art   

                                   lemma  
0                Toronto Dominion Centre  
2           kompleks handlowo-kulturalny  
3                             Joe Fafard  
5               Ludwig Mies van der Rohe  
7  Toronto Dominion Gallery of Inuit Art

In [6]:
print(f"Shape of the Poleval 2019 dataset (only multi word expressions): {mwe_df.shape}")

Shape of the Poleval 2019 dataset (only multi word expressions): (9718, 4)


**remove not polish MWE**


In [7]:
def is_polish_mwe(text):
    """
    Check if the text is a valid Polish MWE.
    Returns True if at least one word is recognized (not tagged as 'ign').
    """
    analysis = morfeusz.analyse(text)
    for _, _, interpretations in analysis:
        tag = interpretations[2]
        if tag != 'ign':
            return True
    return False

In [8]:
df_unrecognized = mwe_df[~mwe_df['text'].apply(is_polish_mwe)].reset_index(drop=True)
print(f"Number of unrecognized MWEs: {len(df_unrecognized)}")
df_unrecognized.sample(10)

Number of unrecognized MWEs: 566


phrase_id  doc_id                                text  \
16       99721  100524             Barvicha Luxury Village   
393     287881  102435                    History Carnival   
482     471925  103954                         Angry Birds   
225      72068  101751                    Creative Commons   
473     471465  103927                       Andrei Strbik   
161      67462  101384  Tiengiz Grigorjewicz Sułakwielidze   
549     490467  107326                   Nicolasem Sarkozy   
63       52562  101186                  Pearson Television   
271      73836  101886                         Two Harbors   
57       52395  101173                        Olle Hansson   

                                  lemma  
16              Barvicha Luxury Village  
393                    History Carnival  
482                         Angry Birds  
225                    Creative Commons  
473                       Andrea Strbik  
161  Tiengiz Grigorjewicz Sułakwielidze  
549                     Nicolas Sarkozy  
63                   Pearson Television  
271                         Two Harbors  
57                         Olle Hansson

In [9]:
mwe_df = mwe_df[mwe_df["text"].apply(is_polish_mwe)].reset_index(drop=True)
print(f"Shape of the Poleval 2019 dataset (only valid Polish MWEs): {mwe_df.shape}")

Shape of the Poleval 2019 dataset (only valid Polish MWEs): (9152, 4)


There was 566 fully not recognized expressions by Morfeusz.


**Remove duplicates**


In [10]:
def remove_duplicates(df):
    """
    Remove duplicates from the DataFrame based on 'text' and 'lemma'.
    Keeps the first occurrence.
    """
    return df.drop_duplicates(subset=["text", "lemma"], keep='first')

In [11]:
mwe_df = remove_duplicates(mwe_df)
print(f"Shape of the Poleval 2019 dataset (after removing duplicates): {mwe_df.shape}")

Shape of the Poleval 2019 dataset (after removing duplicates): (7453, 4)


**TO DO: podzielić na nieodmienne i mianowniki:**


In [12]:
mwe_df[mwe_df["text"] == mwe_df["lemma"]]


phrase_id  doc_id                                   text  \
0         31822   99883                Toronto Dominion Centre   
1        343873   99883           kompleks handlowo-kulturalny   
2         31833   99883                             Joe Fafard   
4        343871   99883  Toronto Dominion Gallery of Inuit Art   
5        343875  100499                        PZL.13 (PZL-13)   
...         ...     ...                                    ...   
9137     329445  107360                           Marin Leovac   
9139     329446  107360                            Roland Linz   
9142     329448  107360                              Tomas Jun   
9144     329449  107360                        Patrick Salomon   
9145     329435  107360                           Bartosz Flis   

                                      lemma  
0                   Toronto Dominion Centre  
1              kompleks handlowo-kulturalny  
2                                Joe Fafard  
4     Toronto Dominion Gallery of Inuit Art  
5                           PZL.13 (PZL-13)  
...                                     ...  
9137                           Marin Leovac  
9139                            Roland Linz  
9142                              Tomas Jun  
9144                        Patrick Salomon  
9145                           Bartosz Flis  

[3625 rows x 4 columns]

In [13]:
mwe_df = mwe_df[mwe_df["text"] != mwe_df["lemma"]]


In [14]:
print(f"Shape of the Poleval 2019 dataset (only inflected mwe): {mwe_df.shape}")


Shape of the Poleval 2019 dataset (only inflected mwe): (3828, 4)


In [15]:
mwe_df[['text', 'lemma']].sample(10)

text                               lemma
2283                             Lexusa IS                            Lexus IS
280                    igrzyskach w Nagano                   igrzyska w Nagano
3188        Sesji Sejmu Dzieci i Młodzieży             Sejm Dzieci i Młodzieży
7531                            A Johnsona                           A Johnson
1673          Anny Marii z domu Petardówny        Anna Maria z domu Petardówny
7322                     stanów kwantowych                      stany kwantowe
1746                         Willa Wrighta                         Will Wright
2100                 Osiedle XXV-lecia PRL               osiedle XXV-lecia PRL
6803  Porozumienia Międzyinstytucjonalnego  Porozumienie Międzyinstytucjonalne
1137                  departamencie Loiret                  departament Loiret

## Cleaning

In [16]:
mwe_df.isna().sum()

phrase_id    0
doc_id       0
text         0
lemma        1
dtype: int64

In [17]:
mwe_df[mwe_df["lemma"].isna()]

phrase_id  doc_id               text lemma
1334     307628  101286  San Diego Rockets   NaN

In [ ]:
# I drop it because it's not a polish MWE
mwe_df = mwe_df.dropna()

In [19]:
mwe_df.isna().sum()

phrase_id    0
doc_id       0
text         0
lemma        0
dtype: int64

In [20]:
mwe_df[mwe_df["text"] == "Dolce & Gabbana"]

phrase_id  doc_id             text                lemma
161      99725  100524  Dolce & Gabbana  Dolce &amp; Gabbana

In [21]:
mwe_df[mwe_df["text"] == "Lublin R-VIII a"]

phrase_id  doc_id             text          lemma
734      65875  101200  Lublin R-VIII a  Lublin R-VIII

In [22]:
mwe_df[mwe_df["text"] == "BBC Radio 1's Live Lounge"]

phrase_id  doc_id                       text        lemma
2264     426217  101434  BBC Radio 1's Live Lounge  BBC Radio 1

In [23]:
mwe_df[mwe_df["text"] == "Jammin' CRT.5"]

phrase_id  doc_id           text   lemma
5108     468926  102125  Jammin' CRT.5  Jammin

In [24]:
mwe_df[mwe_df["text"] == "Traxxas Rustler XL-1 '09"]


phrase_id  doc_id                      text                 lemma
6580     469441  102506  Traxxas Rustler XL-1 '09  Traxxas Rustler XL-1

In [25]:
mwe_df[mwe_df["text"] == "Żerań F S O"]


phrase_id  doc_id         text      lemma
6916     350246  102673  Żerań F S O  Żerań FSO

In [26]:
mwe_df.loc[mwe_df["text"] == "Żerań F S O", "text"] = "Żerań FSO"

In [27]:
mwe_df.loc[mwe_df["text"] == "R C S Orzeł", "text"] = "RCS Orzeł"

In [28]:
mwe_df = mwe_df[mwe_df["text"] != "S K M"]
mwe_df = mwe_df[mwe_df["text"] != "R K S"]

In [29]:
mwe_df[mwe_df["lemma"].str.contains("ZTM")]

phrase_id  doc_id   text lemma
6897     352573  102669  Z T M   ZTM

In [30]:
mwe_df[mwe_df["lemma"].str.contains("WKD")]


phrase_id  doc_id   text lemma
6988     352545  102697  W K D   WKD

In [31]:
mwe_df[mwe_df["text"] == "átlátszó ("]

phrase_id  doc_id        text     lemma
8097     481079  103921  átlátszó (  átlátszó

In [32]:
mwe_df[mwe_df["text"] == "Społecznej Rady Konsultacyjnej ds. aktualizacji „Strategii Rozwoju Bydgoszczy do 2015 roku”"]

phrase_id  doc_id  \
5224      79818  102157   

                                                                                             text  \
5224  Społecznej Rady Konsultacyjnej ds. aktualizacji „Strategii Rozwoju Bydgoszczy do 2015 roku”   

                                                                                        lemma  
5224  Społeczna Rada Konsultacyjna ds. aktualizacji„Strategii Rozwoju Bydgoszczy do 2015 roku

In [33]:
mwe_df.loc[mwe_df["text"] == 'Społecznej Rady Konsultacyjnej ds. aktualizacji „Strategii Rozwoju Bydgoszczy do 2015 roku”', "lemma"] = 'Społeczna Rada Konsultacyjna ds. aktualizacji „Strategii Rozwoju Bydgoszczy do 2015 roku"'

In [34]:
mwe_df[mwe_df["text"].str.contains("Fundacji Ośrodka")]

phrase_id  doc_id                        text                      lemma
6812      71729  102602  Fundacji Ośrodka " Karta "  Fundacja Ośrodka "Karta "

In [35]:
mwe_df.loc[mwe_df["text"] == 'Fundacji Ośrodka " Karta "', ["text", "lemma"]] = ['Fundacji Ośrodka "Karta"', 'Fundacja Ośrodka "Karta"']

In [36]:
mwe_df[mwe_df["text"].str.contains("FILLER")] # error during creating dataset

phrase_id  doc_id                        text                lemma
6973     474289  102694   Dworca FILLER Centralnego    Dworzec Centralny
6978     350524  102696  Dworzec FILLER Śródmieście  Dworzec Śródmieście

In [37]:
mwe_df.loc[mwe_df["text"] == 'Dworca FILLER Centralnego', "text"] = 'Dworca Centralnego'
mwe_df.loc[mwe_df["text"] == 'Dworzec FILLER Śródmieście', "text"] = 'Dworzec Śródmieście'

In [38]:
mwe_df[mwe_df["text"].str.contains("Herburt")]

phrase_id  doc_id                     text                     lemma
1202      50764  101263  Jan Herburt (Arłamowski  Jan Herburt (Arłamowski)

In [39]:
mwe_df[mwe_df["text"].str.contains("Naukę")]

phrase_id  doc_id       text     lemma
4075     310468  101854  Naukę 2.0  Nauka2.0

In [40]:
mwe_df.loc[mwe_df["text"] == "Naukę 2.0", "lemma"] = "Nauka 2.0"


In [41]:
mwe_df[mwe_df["text"] == "Sesji Sejmu Dzieci i Młodzieży"]

phrase_id  doc_id                            text  \
3188      63764  101638  Sesji Sejmu Dzieci i Młodzieży   

                        lemma  
3188  Sejm Dzieci i Młodzieży

In [42]:
mwe_df.loc[mwe_df["text"] == "Sesji Sejmu Dzieci i Młodzieży", "lemma"] = "Sesja Sejmu Dzieci i Młodzieży"

In [43]:
mwe_df[mwe_df["text"]=='ustawy o zmianie ustawy Kodeks postępowania karnego']

phrase_id  doc_id                                                 text  \
3279      63911  101644  ustawy o zmianie ustawy Kodeks postępowania karnego   

                                                   lemma  
3279  Ustawao zmianie ustawy Kodeks postępowania karnego

In [44]:
mwe_df.loc[mwe_df["text"]=='ustawy o zmianie ustawy Kodeks postępowania karnego', "lemma"] = 'ustawa o zmianie ustawy Kodeks postępowania karnego'


In [45]:
mwe_df[mwe_df["text"].str.contains("o ujawnieniu pracy lub służby")]


phrase_id  doc_id  \
4449      81044  101946   
4451      81046  101946   

                                                                                                                                                                               text  \
4449                    o zmianie ustawy o ujawnieniu pracy lub służby w organach bezpieczeństwa państwa lub współpracy z nimi w latach 1944–1990 osób pełniących funkcje publiczne   
4451  ustawie z dnia 11 kwietnia 1997 r. o ujawnieniu pracy lub służby w organach bezpieczeństwa państwa lub współpracy z nimi w latach 1944-1990 osób pełniących funkcje publiczne   

                                                                                                                                                                             lemma  
4449                             ustawa o ujawnieniu pracy lub służby w organach bezpieczeństwa państwa lub współpracy z nimi w latach 1944–1990 osób pełniących funkcje publiczne  
4451  Ustawa z dnia 11 kwietnia 1997 r. o ujawnieniu pracy lub służby w organach bezpieczeństwa państwa lub współpracy z nimi w latach 1944-1990 osób pełniących funkcje publiczne

In [46]:
mwe_df = mwe_df.drop(index=4449)

In [47]:
mwe_df.drop(mwe_df[mwe_df["text"].isin([
    "Dolce & Gabbana", # not a polish MWE
    "Lublin R-VIII a", # ???
    "BBC Radio 1's Live Lounge", # not a polish MWE
    "Jammin' CRT.5", # not a polish MWE
    "Traxxas Rustler XL-1 '09", # not a polish MWE
    "átlátszó (", # not a polish MWE
    "Z T M",
    "W K D"
])].index, inplace=True)

In [48]:
print(f"Shape of the Poleval 2019 dataset (after cleaning): {mwe_df.shape}")

Shape of the Poleval 2019 dataset (after cleaning): (3816, 4)


In [49]:
mwe_df.to_csv("mwe_cleaned_df.csv")

## Actual analysis

In [3]:
mwe_df = pd.read_csv("mwe_cleaned_df.csv", index_col=0)
print(f"Shape of the cleaned Poleval 2019 dataset: {mwe_df.shape}")

Shape of the cleaned Poleval 2019 dataset: (3816, 4)


In [4]:
mwe_df.head()

phrase_id  doc_id                          text                      lemma
3      343872   99883     Ludwiga Mies van der Rohe   Ludwig Mies van der Rohe
7      343879  100499       samolotu PZL.23 „Karaś”     samolot PZL.23 „Karaś”
8       31843  100499      Ministerstwa Komunikacji   Ministerstwo Komunikacji
11     343880  100499  silnikiem Pratt-Whitney Wasp  silnik Pratt-Whitney Wasp
14      31839  100499            Stanisława Praussa           Stanisław Prauss

In [5]:
mwe_df.isna().sum()

phrase_id    0
doc_id       0
text         0
lemma        0
dtype: int64

In [6]:
def is_conjunction(word):
    """
    Check if the word is a conjunction.
    Returns True if the word is a conjunction.
    """
    analysis = morfeusz.analyse(word)
    for _, _, interpretations in analysis:
        tag = interpretations[2]
        if tag.startswith('conj'):
            return True
    return False

In [ ]:
def analyze_pair(words, lemmas):
    """
    Analyze a 2 word expression.
    :param words: List of 2 words in the expression.
    :param lemmas: List of 2 lemmas corresponding to the words.
    :return: number, case, relation type and head of the expression.
    """
    # possible numbers and cases for the words
    number_case = {}
    part_of_speech = {}

    # returned values
    number = None
    case = None
    relation = None
    head = None

    #print(f"Analyzing pair: {words} with lemmas: {lemmas}")

    for i, word in enumerate(words):
        number_case_i = set()
        analyses = morfeusz.analyse(word)

        for _, _, interpretations in analyses:
            tag = interpretations[2].split(":")
            if len(tag) > 2:
                number_i = tag[1]
                case_i = tag[2].split('.')

                # Morfeusz can return more than 1 possibilty of case after the dot so I make a list of it 
                for c in case_i:
                    number_case_i.add((number_i, c))

                part_of_speech[word] = tag[0]

        number_case[i] = number_case_i

    available = [i for i in number_case if number_case[i]]

    # check if it is possible to find common number and case for the words in the mwe -> ZWIĄZEK ZGODY
    common_cases = set.intersection(*number_case.values())
    if len(common_cases)>=1:
        # if there is only one common pair(number, case), then it is an agreement relation and the number and case of the expression can be assigned to the whole expression
        
        if len(common_cases) == 1:
            number_case_head_list = list(common_cases)[0]
            number = number_case_head_list[0]
            case = number_case_head_list[1]
 
        else:
            number = [x[0] for x in list(common_cases)]
            case = [x[1] for x in list(common_cases)]
            
        relation = "agreement"

        # noun -> HEAD
        for i, (word, tag) in enumerate(part_of_speech.items()):
            if tag == 'subst':
                head = lemmas[i]
                break

        # if there is no noun in the expression, then the head is the first word in the expression
            else:
                head = lemmas[0]

    elif [w_i for w_i, l_i in zip(words, lemmas) if w_i == l_i]:
        # government relation    
        for i, word in enumerate(words):
            if word == lemmas[i]:
                continue
            else: 
                if number_case[i]:
                    head = lemmas[i]
                    number_case_head_list = list(number_case[i])
                    number = [x[0] for x in number_case_head_list]
                    case = [x[1] for x in number_case_head_list]
                    relation = "government"
        
    if len(available) == 1:
        #print(f"Only one word has a valid analysis: {words[available[0]]} with lemma {lemmas[available[0]]}, words: {words}, lemmas: {lemmas}, number_case: {number_case}")
        # if there is only one word in the expression that has a valid analysis, then it is the head of the expression
        head = lemmas[available[0]]
        number_case_head_list = list(number_case[available[0]])
        number = [x[0] for x in number_case_head_list]
        case = [x[1] for x in number_case_head_list]
        
        idx_j = 1 - available[0]  # index of the second word in the pair
        if words[idx_j] == lemmas[idx_j]:
            # if the second word is the same as the lemma, then it is a government relation
            relation = "government"
        else:
            relation = "agreement"

    return number, case, relation, head


def analyze_mwe(phrase, lemma):
    """
    Analyze a multi-word expression.
    :param phrase: The multi-word expression.
    :param lemma: The lemma of the multi-word expression.
    :return: analysis results - possible: numbers, cases, relations, heads, pairs.
    """

    phrase = phrase.replace(" - ", "-")
    phrase = phrase.replace("___", "")

    
    words = phrase.strip().split()
    lemmas = lemma.strip().split() 

    numbers = []
    cases = []
    relations = []
    heads = []
    pairs = []

    i = 0
    while i < len(words)-1:
        if is_conjunction(words[i+1]) and i + 2 < len(words):
            pair_words = [words[i], words[i+2]]
            pair_lemmas = [lemmas[i], lemmas[i+2]]
            i += 2 # skip the conjunction
        
        else:
            pair_words = words[i:i+2]
            pair_lemmas = lemmas[i:i+2]
            i += 1 

        # analyze the pair of words
        analyses = analyze_pair(pair_words, pair_lemmas)
        number, case, relation, head = analyses
        pairs.append(tuple(pair_words))
        numbers.append(number)  
        cases.append(case)
        relations.append(relation)
        heads.append(head)


    return numbers, cases, relations, heads, pairs

In [8]:
mwe_df[["numbers", "cases", "relations", "heads", "pairs"]] = mwe_df.apply(lambda row: pd.Series(analyze_mwe(row["text"], row["lemma"])), axis=1)

In [32]:
mwe_df.sample(20)

phrase_id  doc_id                                       text  \
7307     468088  103621                     atomów trójpoziomowych   
5200     468117  102155                         twierdzenia Eulera   
5388     467703  102205                         biegłość żeglarską   
421       40626  100626                          Sokolnikach Lesie   
932      352109  101232                   armią Republiki Wietnamu   
7876     484379  103861                            E. Benveniste’a   
874       53788  101224                    Danieli z Kleinbergerów   
4421      78544  101943  Europejskiemu Inspektorowi Ochrony Danych   
5921     484835  102378                     bibliotekami cyfrowymi   
3753     472766  101745                          centrum handlowym   
7146     466632  102770                      spadochronie Mars 330   
6723     466412  102561             inwestycji infrastrukturalnych   
589      310423  101179                      Lubelskiego Lipca '80   
1109      50470  101250                            Bolesława Prusa   
1841      55185  101361                              T. Kościuszki   
8434     486154  107265               koszykarza Dariusa Lukminasa   
2365     427617  101449                       księcia Ziemowita IV   
4682     349218  102030       Zespołu Szkół nr 2 im. Emilii Plater   
7901     484427  103866                  ośrodków układu nerwowego   
447      420325  100630                       Ekologiem Wojsławice   

                                      lemma  \
7307                    atomy trójpoziomowe   
5200                     twierdzenie Eulera   
5388                     biegłość żeglarska   
421                           Sokolniki Las   
932                armia Republiki Wietnamu   
7876                          E. Benveniste   
874                 Daniela z Kleinbergerów   
4421    Europejski Inspektor Ochrony Danych   
5921                     biblioteki cyfrowe   
3753                       centrum handlowe   
7146                    spadochron Mars 330   
6723           inwestycje infrastrukturalne   
589                     Lubelski Lipiec '80   
1109                          Bolesław Prus   
1841                          T. Kościuszko   
8434              koszykarz Darius Lukminas   
2365                     książę Ziemowit IV   
4682  Zespół Szkół nr 2 im.   Emilii Plater   
7901               ośrodki układu nerwowego   
447                       Ekolog Wojsławice   

                                                                    numbers  \
7307                                                                   [pl]   
5200                                                                   [sg]   
5388                                                                   [sg]   
421                                                                  [None]   
932                                                              [[sg], sg]   
7876                                                             [[sg, sg]]   
874                     [[sg, pl, sg, pl, sg, sg, sg, pl, sg], [gen, inst]]   
4421                                                         [sg, [sg], pl]   
5921                                                                   [pl]   
3753                                                             [[sg, sg]]   
7146  [[sg, sg], [sg, sg.pl, sg.pl, sg.pl, sg, sg.pl, sg.pl, sg.pl, sg.pl]]   
6723                                                                   [pl]   
589                                                    [[sg, sg], [sg, sg]]   
1109                                                             [[sg, sg]]   
1841                                                     [[pl, pl, sg, pl]]   
8434                                                       [[sg, sg], None]   
2365                                                   [[sg, sg], [sg, sg]]   
4682                               [[sg], None, [sg, sg], [pl], None, None]   
7901                                          

In [34]:
mwe_df.sample(20)

phrase_id  doc_id                                text  \
7084     466566  102756  Departamentowi Pożytku Publicznego   
5814     337526  102353                    Zbigniewa Ziobro   
6982     351874  102697                Warszawy Śródmieście   
3516     467159  101678              premiera Donalda Tuska   
972       54168  101236          Bolesława III Krzywoustego   
8587     318908  107287                 Wszystkich Świętych   
948      308653  101234                Chrystusa Zbawiciela   
6987     351877  102697                  Dworca Śródmieście   
9035     327777  107354                     Woli Mieleckiej   
5366     467549  102181               Ignacym Wołodkowiczem   
4801     484640  102050                        gminy Zgierz   
442       40731  100628        Komendzie Powiatowej Policji   
588       52475  101179         Ronda Lubelskiego Lipca '80   
5608     466057  102235                       Kodeksu pracy   
5249      79845  102159                 Prymasa Tysiąclecia   
6946     474222  102684               transportu miejskiego   
1067     307295  101247              Sly & the Family Stone   
1301      51048  101280                   Ameryki Północnej   
3512     467148  101677                   ustawie rowerowej   
543       99808  101171                         Nowej Hucie   

                                lemma  \
7084  Departament Pożytku Publicznego   
5814                  Zbigniew Ziobro   
6982             Warszawa Śródmieście   
3516              premier Donald Tusk   
972           Bolesław III Krzywousty   
8587                   Wszyscy Święci   
948                Chrystus Zbawiciel   
6987              dworzec Śródmieście   
9035                    Wola Mielecka   
5366               Ignacy Wołodkowicz   
4801                     gmina Zgierz   
442         Komenda Powiatowa Policji   
588       Rondo Lubelskiego Lipca '80   
5608                     Kodeks pracy   
5249               Prymas Tysiąclecia   
6946                transport miejski   
1067       Sly &amp; the Family Stone   
1301                 Ameryka Północna   
3512                  ustawa rowerowa   
543                         Nowa Huta   

                                                                  numbers  \
7084                                                   [[pl, pl, sg], sg]   
5814                                                                 [sg]   
6982                                                   [[pl, pl, sg, pl]]   
3516                                                 [[sg, sg], [sg, sg]]   
972                                              [[sg, sg, sg], [sg, sg]]   
8587                                                       [[pl, pl, pl]]   
948                                                            [[sg, sg]]   
6987                                                               [[sg]]   
9035                                                       [[sg, sg, sg]]   
5366                                                                 [sg]   
4801                                                   [[pl, pl, sg, pl]]   
442                                              [[sg, sg], [sg, sg, sg]]   
588                                              [sg, [sg, sg], [sg, sg]]   
5608                                                                 [sg]   
5249                                                                 [sg]   
6946                                                                 [sg]   
1067  [None, None, [sg, sg.pl, sg.pl, sg.pl, sg.pl, sg.pl, sg.pl, sg.pl]]   
1301                                                                 [sg]   
3512                                                           [[sg, sg]]   
543                                                            [[sg, sg]]   

                                                        cases  \
7084                                   [[voc, nom, dat], gen]   
5814                                                    [acc]   
6982   

In [17]:
analyze_pair(["Objawienia",  "Pańskiego"], ["Objawienie", "Pański"])	

('sg', 'gen', 'agreement', 'Pański')

In [22]:
morfeusz.analyse("diecezji koszalińsko-kołobrzeskiej")

[(0, 1, ('diecezji', 'diecezja', 'subst:sg:gen:f', ['nazwa_pospolita'], [])),
 (0,
  1,
  ('diecezji', 'diecezja', 'subst:sg:dat.loc:f', ['nazwa_pospolita'], [])),
 (0,
  1,
  ('diecezji', 'diecezja', 'subst:pl:gen:f', ['nazwa_pospolita'], ['hom.'])),
 (1, 2, ('koszalińsko', 'koszaliński', 'adja', [], [])),
 (2, 3, ('-', '-', 'interp', [], [])),
 (3, 4, ('kołobrzeskiej', 'kołobrzeski', 'adj:sg:dat:f:pos', [], [])),
 (3, 4, ('kołobrzeskiej', 'kołobrzeski', 'adj:sg:gen:f:pos', [], [])),
 (3, 4, ('kołobrzeskiej', 'kołobrzeski', 'adj:sg:loc:f:pos', [], []))]